main/CTLIterator.ml

```diff
@@ -0,0 +1,549 @@
+open AbstractSyntax
+open CTLProperty
+open Apron
+open Domain
+open Partition
+open Functions
+open Iterator
+
+(* type for CTL properties, instantiated with bExp for atomic propositions *)
+type ctl_property = AbstractSyntax.bExp CTLProperty.generic_property
+
+type quantifier = UNIVERSAL | EXISTENTIAL
+
+let rec print_ctl_property fmt (property:ctl_property) = match property with 
+  | Atomic b -> AbstractSyntax.bExp_print_aux fmt b
+  | AX p -> Format.fprintf fmt "AX{%a}" print_ctl_property p
+  | AF p -> Format.fprintf fmt "AF{%a}" print_ctl_property p
+  | AG p -> Format.fprintf fmt "AG{%a}" print_ctl_property p
+  | AU (p1,p2) -> Format.fprintf fmt "AU{%a}{%a}" print_ctl_property p1 print_ctl_property p2;
+  | EX p -> Format.fprintf fmt "EX{%a}" print_ctl_property p
+  | EF p -> Format.fprintf fmt "EF{%a}" print_ctl_property p
+  | EG p -> Format.fprintf fmt "EG{%a}" print_ctl_property p
+  | EU (p1,p2) -> Format.fprintf fmt "EU{%a}{%a}" print_ctl_property p1 print_ctl_property p2;
+  | AND (p1,p2) -> Format.fprintf fmt "AND{%a}{%a}" print_ctl_property p1 print_ctl_property p2;
+  | OR (p1,p2) -> Format.fprintf fmt "OR{%a}{%a}" print_ctl_property p1 print_ctl_property p2;
+  | NOT p -> Format.fprintf fmt "NOT{%a}" print_ctl_property p;
+
+
+module InvMap = Map.Make(struct type t=label let compare=compare end)
+
+
+(* Bundle commonly used values (AST, Apron env. variable list) to one struct *)
+type program = {
+  environment: Apron.Environment.t;
+  variables: AbstractSyntax.var list;
+  mainFunction: AbstractSyntax.func;
+  globalBlock: AbstractSyntax.block; 
+  dummyPosition: IntermediateSyntax.extent;
+}
+
+(* Computes the set of all labels of a program *)
+let labels_of_program program = 
+  let rec stmtLabels s =
+    match s with
+    | A_if (_, s1, s2) -> List.append (blockLabels s1) (blockLabels s2)
+    | A_while (l,(b,ba),s) -> l::(blockLabels s)
+    | _ -> []
+  and blockLabels b =
+    match b with
+    | A_empty l -> [l]
+    | A_block (l,(s,_),b) -> l::(List.append (stmtLabels s) (blockLabels b))
+  in
+  (blockLabels program.globalBlock) @ (blockLabels program.mainFunction.funcBody) 
+
+
+(* get label at start of block *)
+let block_label block = 
+  match block with
+  | A_empty l -> l
+  | A_block (l,_,_) -> l
+
+
+
+(*
+   This function takes a given paresed program and introduces a new variable called 'termination' that is set to '0'
+   whenever the program terminates i.e. at the end and before every return statement. 
+
+   The augmented program and a corresponding ctl_property for termination is returned. 
+   This pair can then be used with the usual CTL analysis to check for termination.
+*)
+let program_of_prog_with_termination (prog: AbstractSyntax.prog) (main: AbstractSyntax.StringMap.key) : (program * ctl_property) =
```

Comment:
That's a good idea, would you suggest augmenting the AST by adding a final label before each 'return' statement and at the end? 
We could call the label 'exit' and check for termination with 'AF exit:true'